incubator/hnc/pkg/validators/config.go

```diff
@@ -0,0 +1,182 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+)
+
+// ConfigServingPath is where the validator will run. Must be kept in sync with the
+// kubebuilder markers below.
+const (
+	ConfigServingPath = "/validate-hnc-x-k8s-io-v1alpha1-hncconfigurations"
+)
+
+// Note: the validating webhook FAILS OPEN. This means that if the webhook goes down, all further
+// changes to the objects are allowed.
+//
+// +kubebuilder:webhook:path=/validate-hnc-x-k8s-io-v1alpha1-hncconfigurations,mutating=false,failurePolicy=fail,groups="hnc.x-k8s.io",resources=hncconfigurations,verbs=create;update;delete,versions=v1alpha1,name=hncconfigurations.hnc.x-k8s.io
+
+type Config struct {
+	Log     logr.Logger
+	client  validationClient
+	decoder *admission.Decoder
+}
+
+// validationClient checks if a resource exists. The check should typically be performed against the apiserver,
+// but need to be stubbed out during unit testing.
+type validationClient interface {
+	// ResourceExists takes a GVK, and returns true if the corresponding resource of the GVK exists in the apiserver.
+	ResourceExists(ctx context.Context, gvk schema.GroupVersionKind) (bool, error)
+}
+
+type set map[schema.GroupVersionKind]bool
+
+func (c *Config) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		return allow("HNC SA")
+	}
+
+	if req.Operation == v1beta1.Delete {
+		if req.Name == "config" {
+			return deny(metav1.StatusReasonForbidden, "Deleting the 'config' object is forbidden")
+		} else {
+			// We allow deleting other objects. If the validation controller has always been running, we should not
+			// enter this case because objects of other names cannot be created. If users somehow bypass the
+			// validation controller and create objects of other names, we will allow them to delete the objects.
+			return allow("")
+		}
+	}
+
+	inst := &api.HNCConfiguration{}
+	if err := c.decoder.Decode(req, inst); err != nil {
+		c.Log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	resp := c.handle(ctx, inst)
+	c.Log.Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the validation logic of this validator for Create and Update operations,
+// allowing it to be more easily unit tested (ie without constructing a full admission.Request).
+func (c *Config) handle(ctx context.Context, inst *api.HNCConfiguration) admission.Response {
+	if inst.GetName() != "config" {
+		return deny(metav1.StatusReasonForbidden, fmt.Sprintf("Wrong singleton name: %s; the name should be 'config'", inst.GetName()))
```

Comment:
StatusReasonBadRequest? Or maybe StatusReasonInvalid?

Reply:
Thanks for the suggestion! Looks like `StatusReasonInvalid` is more appropriate. Changed to that.